### PR TITLE
Re-enable Pyodide CI job and testing for SymPy

### LINKED ISSUE #24306: Pyodide tests failing when invoked with pytest
This is specifically related to an issue associated with PR #24291, which is aiming to move SymPy's default test runner to pytest.

I'm struggling to get the Pyodide tests to pass in CI. An example of the failure can be found [here](https://github.com/sympy/sympy/actions/runs/3542327630/jobs/5947722359) (for the time being). The issue appears to be that pytest is not able to discover any tests. Any ideas on how to attack this would be greatly appreciated.

@eagleoflqj, perhaps you can help as the author of `bin/test_pyodide.mjs`?

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
#### References to other Issues or PRs

This PR is based on conversations with @oscarbenjamin in https://github.com/pyodide/pyodide/pull/5098 and a subsequent discussion with @asmeurer a fortnight ago. The Pyodide test suite was last discussed in #24920 and https://github.com/sympy/sympy/pull/25394#issuecomment-1650410409. 

Furthermore, this PR can potentially close #24306.

#### Brief description of what is fixed or changed

This PR aims to re-enable the Pyodide CI job that runs the SymPy test suite inside a Pyodide runtime and fix some of the issues with reliability and flaky tests. I've removed `bin/test_pyodide.mjs` and the related code for it, and removed imports for `pyodide_js` across the codebase where necessary. The rest of the changes are just related to code cleanup and adding skip markers for a few tests. The Pyodide virtual environment that has been added should be more closer to a Pythonic test invocation and more useful if wanting to run the tests locally. While a GHA runner will be slower, I'm able to run the test suite minus the slow and "tooslow" tests without hiccups in thirty minutes when testing locally (side note: the "tooslow" tests take way too long in Pyodide, up to the order of eight minutes for a single test – they should definitely be disabled!). Additionally, the `@skip_under_pyodide` decorator added in #24892 has been simplified.

SymPy is a pure Python package and testing it under Pyodide may not have been useful, but since a CI job previously existed and if the test suite can run reliably here, both of those points make an additional case for including it back again :D 

P.S. It's always good to have a CI job in general (as long as it doesn't strain SymPy's existing infrastructure) because there are some "slow" tests that I noted were failing and have been previously noted to fail (which I haven't tried to fix here – I could try them in a separate PR after this since they might be tricky).

#### Other comments

The tests are split into a GitHub Actions matrix with four groups with `pytest-split`, as it was in the previous iteration of the test suite before it was disabled. There is a caching mechanism where the test durations are individually collected and stored in a GitHub Actions cache with a persistent key, so that it is always saved and restored every time the workflow is triggered. This seems to work quite well when a cache hits (which will be every time) when I was testing this in a PR on my fork (https://github.com/agriyakhetarpal/sympy/pull/1), but I can let go of this if needed and switch to the existing periodic update for the durations (the durations for Pyodide will be stored and saved separately, however).

Due to these additional jobs, I've added all this under a separate workflow in `emscripten.yml` and not in the standard `runtests.yml` file (but I'm happy to move things around if I'm requested to). 

With the caching here, all four test groups complete in ~19 minutes.

#### Release Notes

NO ENTRY

(? I'm unsure, please let me know if this PR needs one 😅)

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Once I confirm that everything works, I can rewrite my commit history here in adherence to https://docs.sympy.org/latest/contributing/new-contributors-guide/workflow-process.html.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:38]: I have plans to make this easier with https://github.com/pyodide/pyodide-actions/issues/12 which will enable Dependabot updates for these versions and delegate the update procedure to us.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:5]:
```suggestion
name: Pyodide CI
```

based on discussions in https://github.com/sympy/sympy/issues/24920. Should I make the job name shorter below?

**Comment by agriyakhetarpal** [comment]
The [first workflow run](https://github.com/sympy/sympy/actions/runs/11454441036?pr=27183) is supposed to take longer since the caches have yet to be stashed. Essentially, there isn't a neat solution in https://github.com/jerry-git/pytest-split/issues/20 and https://github.com/jerry-git/pytest-split/pull/91 hasn't been merged/made it to a release yet, so what I've done here is implemented my own.

**Comment by agriyakhetarpal** [comment]
The [first workflow run](https://github.com/sympy/sympy/actions/runs/11454441036?pr=27183) is supposed to take longer since the caches have yet to be stashed. Essentially, there isn't a neat solution in https://github.com/jerry-git/pytest-split/issues/20 and https://github.com/jerry-git/pytest-split/pull/91 hasn't been merged/made it to a release yet, so what I've done here is implemented my own (which isn't too invasive and, IMO, quite easy to comprehend and maintain).

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: Addressed in 0d81b906191caacb87deb7dc614d036d3b78f805.

**Comment by oscarbenjamin** [comment]
CC @eagleoflqj 

This is great. Thanks @agriyakhetarpal!

The changes here look good to me. I'll give some time for others to review this but otherwise I think this is good to merge.

For background the pyodide CI jobs were previously removed because they were flakey. In particular the tests would pass but then pyodide would sometimes hang instead of exiting. I'm not sure what caused that but this CI job runs pyodide in a different way which will hopefully not have that problem.

**Review by anutosh491 — APPROVED** [review]
Thanks for this @agriyakhetarpal 
Other than squashing the commits into some meaningful ones, this looks good !

**Comment by oscarbenjamin** [comment]
> I need help finding a place where SymPy uploads these wheels.

These are not uploaded anywhere because SymPy is easy to build and install:
```
pip install git+https://github.com/sympy/sympy.git@master
```
Maybe it would be better to have wheels though for something like SymPy Live and perhaps some users would be happier using nightly wheels rather than installing from git (even though it basically gives the same thing).

It would be good to upload them to the Anaconda nightly wheels since that is where most of SymPy's dependencies put them. I guess we need to ask someone for permission to upload there?

> the SymPy version would always stay up to date and not be tied in with the Pyodide version

Yes, that would be better.

Also for SymPy Live we should install python-flint as well as SymPy. SymPy 1.13.1 can use python-flint 0.6.0 to make some things faster:
```console
$ isympy
IPython console for SymPy 1.14.dev (Python 3.12.2-64-bit) (ground types: flint)
...
In [1]: import sympy

In [3]: type(sympy.ZZ(1))
Out[3]: flint.types.fmpz.fmpz
```
There are builds of python-flint available in pyodide and typing `import flint` in jupyterlite installs python-flint. We need it to be available before `import sympy` runs though because SymPy decides at import time whether or not to use `GROUND_TYPES=flint` based on whether the module is available.

**Comment by oscarbenjamin** [comment]
> ```
> $ isympy
> IPython console for SymPy 1.14.dev (Python 3.12.2-64-bit) (ground types: flint)
> ```

The live website should be changed to display this sort of information at startup like isympy does.

**Comment by agriyakhetarpal** [comment]
> These are not uploaded anywhere because SymPy is easy to build and install:

I agree on the ease of installation, though I would say it is still useful to provide wheels. Having a pre-built wheel means that the task of installing it is just copying over files from the archive to the installation directory, but installing from a GitHub URL means that the build frontend has to be invoked (which, in-turn, invokes the build backend – `setuptools` here) to build a wheel, which isn't the best thing to do in terms of security (all the more vital in a web browser!) because building a wheel from the source requires setup code to be executed. `pyodide-build` has been built upon monkeypatches over `pypa/build`, but the output wheel is identical for pure Python projects, as is the case here. So any other standards-compliant build frontend should work, too, such as the new `uv build` offering or the like.

> Maybe it would be better to have wheels though for something like SymPy Live and perhaps some users would be happier using nightly wheels

> It would be good to upload them to the Anaconda nightly wheels since that is where most of SymPy's dependencies put them. I guess we need to ask someone for permission to upload there?

Absolutely, the state where a package has wheels for development builds is much easier to deal with, e.g., for those who would like to test against the SymPy nightlies. Yes, since you're keen, you may ask for permission to upload the nightlies by opening an issue here: https://github.com/scientific-python/upload-nightly-action#access-to-the-scientific-python-nightly-wheels-channel. I imagine SymPy being a big project shouldn't have any issues receiving said access :)

> Also for SymPy Live we should install python-flint as well as SymPy. SymPy 1.13.1 can use python-flint 0.6.0 to make some things faster:
> There are builds of python-flint available in pyodide and typing `import flint` in jupyterlite installs python-flint

Interesting! Yes, in that case, we could either build two wheels, one for SymPy and another for `python-flint`. Is `python-flint` updated regularly? As far as https://pypi.org/project/python-flint/#history is concerned (based on the non-pre-releases), the version in Pyodide should suffice. However, if we are behind the latest available version for any reason at some point in time and SymPy uses the latest available version, it might make sense to compile both (out-of-tree, that is) and serve them in the live site.

> > ```
> > $ isympy
> > IPython console for SymPy 1.14.dev (Python 3.12.2-64-bit) (ground types: flint)
> > ```
> 
> The live website should be changed to display this sort of information at startup like isympy does.

Could you share how this has been implemented? I don't know too much about JupyterLite, so I'd like to explore whether this is possible within the kernel. I checked https://github.com/sympy/sympy/blob/master/isympy.py, but I don't see any references to `flint` in there.

**Comment by oscarbenjamin** [comment]
I've just asked permission to upload nightly wheels for SymPy and python-flint in https://github.com/scientific-python/upload-nightly-action/issues/111

**Comment by agriyakhetarpal** [comment]
The test failure logs look like I missed something when rebasing, I'll take a brief moment to check.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:9]:
```suggestion
on: [push, pull_request]
```

**Comment by oscarbenjamin** [comment]
> Is `python-flint` updated regularly?

Yes. I am currently holding off on putting out the next python-flint release. There are a lot of new features since 0.6.0 and I am hoping that we can test sympy integration for some of them before putting out a final release in case some changes are needed in python-flint. I believe that @Jake-Moss is working on multivariate polynomial integration (gh-26793) after adding those features in python-flint. Once we have at least some of that there will be a 0.7.0 release of python-flint but I don't want to make a final release yet if there is a possibility that we might want to change any of the new interfaces.

> > > ```
> > > $ isympy
> > > IPython console for SymPy 1.14.dev (Python 3.12.2-64-bit) (ground types: flint)
> > > ```
> > 
> > 
> > The live website should be changed to display this sort of information at startup like isympy does.
> 
> 
> Could you share how this has been implemented?

The flint part comes from `GROUND_TYPES`:
https://github.com/sympy/sympy/blob/d7e4c07d08f9e86d483b6e28fdb04f3cabfe70e4/sympy/external/gmpy.py#L109-L116
If the ground types are flint then you should be able to import flint and find the version:
```python
In [29]: from sympy.external.gmpy import GROUND_TYPES

In [30]: GROUND_TYPES
Out[30]: 'flint'

In [31]: import flint

In [32]: flint.__version__
Out[32]: '0.7.0a3'
```
The information that I think should be displayed is:

- relevant versions for pyodide/jupyterlite
- version of mpmath
- version of sympy
- GROUND TYPES
- version of python-flint if ground types are flint.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:74]:
```suggestion
```

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:87]:
```suggestion
```

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:100]:
```suggestion
        run: mv .test_durations test_durations_group_${{ matrix.group }}.json
```

**Comment by agriyakhetarpal** [comment]
One more thing I wish to explore here before this is merged is to examine ways for updating the cache – since there is currently a [warning](https://github.com/sympy/sympy/actions/runs/11463946984/job/31900105689?pr=27183#annotation:5:11) that I didn't notice previously about the fact that one can't overwrite an existing cache. This feature has been open for years at this point: https://github.com/actions/toolkit/issues/505, https://github.com/actions/cache/issues/342, https://github.com/actions/cache/issues/171, https://github.com/actions/cache/issues/109#issuecomment-558771281, and many more. I'll probably use the [`gh` CLI](https://github.com/azu/github-actions-overwrite-cache-example) to overwrite the cache and see if that works.

**Comment by agriyakhetarpal** [comment]
I've fallen into a rabbit hole where the cache deletion works, but only in limited scenarios – it doesn't seem to work here because of how GitHub authentications/permissions work. This PR to SymPy is coming from my fork. It doesn't work on the `push:` event due to the same reason (it doesn't work on `push:` events in general because I've supplied a branch to delete caches from, which would exist only with PRs), but it works on my PR within a fork, of course (https://github.com/agriyakhetarpal/sympy/pull/1).

I'll try to make the cache deletions more fine-grained, I don't think it should add any additional complexity. However, any feedback on the approach is welcome.

I could:
- either remove the `-B` argument and not specify a branch to delete caches from (allowing deletions , or
- keep the argument and add a bunch of if-else clauses to delete scoped caches from a branch if the PR is within a fork (not a common use case) or just fail gracefully, since only a PR from an account with "write" access to the repository can delete the central caches

(There could be scenarios that I might be missing).

**Comment by oscarbenjamin** [comment]
> I could:
> 
> * either remove the `-B` argument and not specify a branch to delete caches from (allowing deletions , or

That sounds reasonable to me. It is just a cache so it should be okay to delete. I don't know if we use the cache for anything else (maybe hypothesis?).

**Comment by oscarbenjamin** [comment]
There are now nightly wheels for both SymPy and python-flint at the Anaconda scientific nightly channel:
https://anaconda.org/scientific-python-nightly-wheels/repo

What is not there is a wasm wheel for python-flint though (https://github.com/flintlib/python-flint/issues/234).

**Comment by agriyakhetarpal** [comment]
Thanks for the feedback. I'm here now after a conference and vacation, I'll wrap this PR up for merging within the day.

**Comment by agriyakhetarpal** [comment]
I've now added the `|| exit true` statement to the `gh cache` command here to fail more gracefully. With this, the idea is that while external contributors and those without "write" access to the SymPy repository (e.g., me) won't be able to delete caches via PRs, those with write access such as core developers or maintainers should be able to do so via their PRs, regardless of whether the PRs come from their forks of SymPy or from branches within the main SymPy repository. This also applies to PRs by GitHub apps with write access, i.e., Dependabot, or pre-commit, or similar – thus, the caches will keep getting created and saved in case the repository ever becomes inactive for a period of time for any reason.

I'll push another (empty) commit to exercise the cache saves/restores now, and this should be good to go, @oscarbenjamin and @anutosh491! Please let me know if you'd like me to clean up my commit history any further.

**Comment by agriyakhetarpal** [comment]
Also, if we don't wish to deal with GHA caches at all, https://nightly.link/ exists, too, which would be operated solely through GHA artifacts.

File: .github/workflows/emscripten.yml
Comment (oscarbenjamin) [inline .github/workflows/emscripten.yml:131]: We rarely have PRs from branches in the main repository. Although possibly dependabot is an exception...

**Comment by oscarbenjamin** [comment]
I think this looks good but will wait and see if others review.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: Sorry, I noticed that this doesn't include what I mentioned above in https://github.com/sympy/sympy/pull/27183#issuecomment-2457239924 – this is also valid for PRs from maintainers/core developers Updated the comment to match with my statement in 43a123cd6e6de8391b53a8724dbcb19eee3f6834.

File: .github/workflows/emscripten.yml
Comment (oscarbenjamin) [inline .github/workflows/emscripten.yml:40]: We would not normally pin these so tightly but rather just use `@v4` so it doesn't need to be updated so often.

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: Yes, makes sense to me. Addressed in 0040f74652b61482576d416ecd8b7ab52adc9b55.

**Comment by agriyakhetarpal** [comment]
Also, I'll recommend that the durations cache under https://github.com/sympy/sympy/actions/caches be deleted at the time or after this PR gets merged, so that we can start afresh – the central cache is scoped differently from the PR caches, so we won't have accurate results until then.

**Comment by oscarbenjamin** [comment]
Okay let's get this in.

**Comment by oscarbenjamin** [comment]
> I'll recommend that the durations cache under https://github.com/sympy/sympy/actions/caches be deleted at the time or after this PR gets merged

I don't see anything there. Do we need to do anything?

**Comment by agriyakhetarpal** [comment]
> I don't see anything there. Do we need to do anything?

I guess the caches were evicted because it has been greater than one week; i.e., the retention period. We should all be set, thanks for the review and merge!

**Comment by agriyakhetarpal** [comment]
I'll open a follow-up PR to upload these wheels to the Anaconda.org index that was set up in scientific-python/upload-nightly-action#111.

**Comment by oscarbenjamin** [comment]
The merge commit b7aa6e914f157e0b0424b2d8acd038ccd6178223 has failed CI checks but I merged another PR with commit 353047ceb08cf57c2412d06fb8d76cf79e9f7583 that seems to be running okay...

**Comment by oscarbenjamin** [comment]
> I'll open a follow-up PR to upload these wheels to the Anaconda.org index

I've already done that for SymPy in gh-27186:
https://anaconda.org/scientific-python-nightly-wheels/sympy

Not yet for python-flint: https://github.com/flintlib/python-flint/issues/234

**Comment by oscarbenjamin** [comment]
> Not yet for python-flint: [flintlib/python-flint#234](https://github.com/flintlib/python-flint/issues/234)

I misstated here: I have setup both sympy and python-flint to upload nightly wheels. What I haven't done is made a nightly WASM build of python-flint.

**Comment by agriyakhetarpal** [comment]
> I misstated here: I have setup both sympy and python-flint to upload nightly wheels. What I haven't done is made a nightly WASM build of python-flint.

Superb, thanks for that! Yes, I'll see when I can find time to test out a WASM build for `python-flint` locally – we can continue our exchange on https://github.com/flintlib/python-flint/issues/234 for this.

> The merge commit [b7aa6e9](https://github.com/sympy/sympy/commit/b7aa6e914f157e0b0424b2d8acd038ccd6178223) has failed CI checks but I merged another PR with commit [353047c](https://github.com/sympy/sympy/commit/353047ceb08cf57c2412d06fb8d76cf79e9f7583) that seems to be running okay...

I'll take a look! Not exactly sure why it failed there, perhaps I missed something.

**Comment by oscarbenjamin** [comment]
> commit [353047c](https://github.com/sympy/sympy/commit/353047ceb08cf57c2412d06fb8d76cf79e9f7583) that seems to be running okay...

It looks like the relevant CI jobs have passed there now. We should keep an eye on this as new PRs are opened etc but I guess it should be fine...

Thanks for all of your help with this @agriyakhetarpal!

**Pull request merged**
